### PR TITLE
Race Condition Fix for not always finding ports for apps

### PR DESCRIPTION
**Pull request by tateexon**
We can sometimes pull the list of apps and ports to check for ready before all the apps and ports are populated. This forces us to wait for the full list of apps added to helm to be present before starting this logic.

### CODE REVIEW

File: client/client.go [hunk 1]
```diff
@@ -255,7 +255,7 @@ type ReadyCheckData struct {
 }
 
 // CheckReady application heath check using ManifestOutputData params
-func (m *K8sClient) CheckReady(namespace string, c *ReadyCheckData) error {
+func (m *K8sClient) CheckReady(namespace string, c *ReadyCheckData, appCount int) error {
 	// wait for the number of enumerated apps to be at least 1 before checking
 	// for ready or we can error out on slow runs or large jobs
```
Comment (skudasov) [inline client/client.go:260]: Please, update the comment.

File: client/client.go [hunk 2]
```diff
@@ -255,7 +255,7 @@ type ReadyCheckData struct {
 }
 
 // CheckReady application heath check using ManifestOutputData params
-func (m *K8sClient) CheckReady(namespace string, c *ReadyCheckData) error {
+func (m *K8sClient) CheckReady(namespace string, c *ReadyCheckData, appCount int) error {
```
Comment (skudasov) [inline client/client.go:258]: We now have 2 checks in `CheckReady`:
- Checking we are seeing all pods with `app` label
- Checking all pods are `Running` and that all containers are `Ready`
Both checks are doing almost the same thing
I think the race is here in `WaitPodsReady` between
```
podList, err := m.ListPods(ns, rcd.ReadinessProbeCheckSelector)
```
And the code that checks that list afterwards, can we, maybe, check that list for pods count, not apps count?

Comment (skudasov) [reply client/client.go:258]: So the problem is that we are not validating list for consistency with our deployment.
We are also checking all the apps with the label `app` are equal to all the pods we've found in replicas, I think it'd break if there will be a single deployment without `app` label? Maybe we can remove that convention now, when we started to check pods number in deployment directly?

Comment (skudasov) [reply client/client.go:258]: Maybe we can just check it as `helm` do
image

Comment (skudasov) [reply client/client.go:258]: In our case `Desired - maxUnavailable` can do the trick, maybe?

Comment (tateexon) [reply client/client.go]: I think that will be a better direction. Will make a proper ticket to fix this a better way. For now we will continue with this to unblock people but we definitely need a better solution. I cleaned up the code a bit and broke it out into logical functions for now.

File: environment/environment.go [hunk 3]
```diff
@@ -467,12 +470,45 @@ func (m *Environment) Deploy(manifest string) error {
...
+							podCount += int(replicas.(float64))
+						} else {
+							podCount += 1
+						}
+					}
+				}
+			}
+		}
+	}
+	config.JSIIGlobalMu.Unlock()
```
[the remark this answers is not in the document]
Comment (tateexon) [reply environment/environment.go]: Good callout. Udated

**Review by kalverra — APPROVED** [review]

File: client/client.go (hunk 2 above)
Comment (skudasov) [reply client/client.go]: https://github.com/helm/helm/blob/76157c6d06f981a42bfbb9ec636bae5443cb2c8f/pkg/kube/ready.go#L91

**Pull request merged**